modules/pricer/src/test/java/com/opengamma/strata/pricer/impl/volatility/smile/SabrHaganNormalVolatilityFormulaTest.java

```diff
@@ -0,0 +1,313 @@
+                    alpha[loopalpha], beta, rho[looprho], nu[loopnu]);
+                double valueExpected = FORMULA.volatility(
+                    forward[loopforward], strikes[loopstrikes], time[looptime],
+                    alpha[loopalpha], beta, rho[looprho], nu[loopnu]);
+                assertThat(resultComputed.getValue()).isCloseTo(valueExpected, TOLERANCE_1);
+                double valueForwardP = FORMULA.volatility(
+                    forward[loopforward] + shiftFd, strikes[loopstrikes], time[looptime],
+                    alpha[loopalpha], beta, rho[looprho], nu[loopnu]);
+                double derivativeForwardExpected = (valueForwardP - valueExpected) / shiftFd;
+                assertThat(resultComputed.getDerivative(0)).isCloseTo(derivativeForwardExpected, TOLERANCE_2);
+                double valueStrikesP = FORMULA.volatility(
+                    forward[loopforward], strikes[loopstrikes] + shiftFd, time[looptime],
+                    alpha[loopalpha], beta, rho[looprho], nu[loopnu]);
+                double derivativeStrikesExpected = (valueStrikesP - valueExpected) / shiftFd;
+                assertThat(resultComputed.getDerivative(1)).isCloseTo(derivativeStrikesExpected, TOLERANCE_2);
+                double valueAlphaP = FORMULA.volatility(
+                    forward[loopforward], strikes[loopstrikes], time[looptime],
+                    alpha[loopalpha] + shiftFd, beta, rho[looprho], nu[loopnu]);
+                double derivativeAlphaExpected = (valueAlphaP - valueExpected) / shiftFd;
+                assertThat(resultComputed.getDerivative(2)).isCloseTo(derivativeAlphaExpected, TOLERANCE_2);
+                double valueBetaP = FORMULA.volatility(
+                    forward[loopforward], strikes[loopstrikes], time[looptime],
+                    alpha[loopalpha], beta + shiftFd, rho[looprho], nu[loopnu]);
+                double derivativeBetaExpected = (valueBetaP - valueExpected) / shiftFd;
+                assertThat(resultComputed.getDerivative(3)).isCloseTo(derivativeBetaExpected, TOLERANCE_2);
+                double valueRhoP = FORMULA.volatility(
+                    forward[loopforward], strikes[loopstrikes], time[looptime],
+                    alpha[loopalpha], beta, rho[looprho] + shiftFd, nu[loopnu]);
+                double derivativeRhoExpected = (valueRhoP - valueExpected) / shiftFd;
+                assertThat(resultComputed.getDerivative(4)).isCloseTo(derivativeRhoExpected, TOLERANCE_1);
+                double valueNuP = FORMULA.volatility(
+                    forward[loopforward], strikes[loopstrikes], time[looptime],
+                    alpha[loopalpha], beta, rho[looprho], nu[loopnu] + shiftFd);
+                double derivativeNuExpected = (valueNuP - valueExpected) / shiftFd;
+                assertThat(resultComputed.getDerivative(5)).isCloseTo(derivativeNuExpected, TOLERANCE_1);
+
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  @Test
```

Comment:
this is only for the beta <> 0.0 case, correct? negative forwards and strikes still work for beta = 0?

Reply:
Yes, for beta=0, negative forward and strikes are good. I'm adding some further tests with negative forwards and strikes in the "Beta0" case.